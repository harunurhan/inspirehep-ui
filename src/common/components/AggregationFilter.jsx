import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Immutable from 'immutable';
import { List, Tag } from 'antd';

import CheckboxItem from './CheckboxItem';

class AggregationFilter extends Component {
  static getDerivedStateFromProps(nextProps, prevState) {
    const selectionMap = nextProps.selectedKeys
      .reduce((map, key) => map.set(key, true), Immutable.Map());

    return {
      ...prevState,
      selectionMap,
    };
  }

  constructor(props) {
    super(props);
    this.state = { selectionMap: Immutable.Map() };
  }

  onBucketChange(key, selected) {
    let { selectionMap } = this.state;
    selectionMap = selectionMap.set(key, selected);
    this.setState({ selectionMap }); // TODO: really needed?
    const selections = selectionMap.keySeq()
      .filter(bucketKey => selectionMap.get(bucketKey))
      .toArray();
    this.props.onChange(selections);
  }

  render() {
    return (
      <div style={{ width: '100%' }}>
        <List header={<strong>{this.props.name}</strong>}>
          {this.props.buckets.map(bucket => (
            <List.Item key={bucket.get('key')}>
              <List.Item.Meta title={
                <CheckboxItem
                  checked={this.state.selectionMap.get(bucket.get('key'))}
                  onChange={(checked) => { this.onBucketChange(bucket.get('key'), checked); }}
                >
                  {bucket.get('key')}
                </CheckboxItem>
              }
              />
              <Tag>
                {bucket.get('doc_count')}
              </Tag>
            </List.Item>
          ))}
        </List>
      </div>
    );
  }
}

AggregationFilter.propTypes = {
  onChange: PropTypes.func.isRequired,
  buckets: PropTypes.instanceOf(Immutable.List).isRequired,
  name: PropTypes.string.isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  selectedKeys: PropTypes.arrayOf(PropTypes.string),
};

AggregationFilter.defaultProps = {
  selectedKeys: [],
};

export default AggregationFilter;
