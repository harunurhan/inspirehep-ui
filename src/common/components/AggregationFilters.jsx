import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Immutable from 'immutable';

import AggregationFilter from '../components/AggregationFilter';

const RANGE_AGGREATION_KEY = 'earliest_date';

class AggregationFilters extends Component {
  static isRange(aggregationKey) {
    return aggregationKey === RANGE_AGGREATION_KEY;
  }

  static compareAggregationEntries([, aggregation1], [, aggregation2]) {
    const order1 = aggregation1.getIn(['meta', 'order']);
    const order2 = aggregation2.getIn(['meta', 'order']);
    return order1 - order2;
  }

  render() {
    const {
      aggregations,
      numberOfResults,
      query,
      onAggregationChange,
    } = this.props;

    return (
      aggregations &&
      numberOfResults > 0 && (
        <div className="bg-white pa3">
          {aggregations
            .entrySeq()
            .filter(([, aggregation]) => aggregation.get('buckets').size > 0)
            .sort(AggregationFilters.compareAggregationEntries)
            .map(([aggregationKey, aggregation]) => (
              <div key={aggregationKey}>
                <AggregationFilter
                  range={aggregationKey === RANGE_AGGREATION_KEY}
                  name={aggregation.getIn(['meta', 'title'])}
                  buckets={aggregation.get('buckets')}
                  selections={query[aggregationKey]}
                  onChange={selections => {
                    onAggregationChange(aggregationKey, selections);
                  }}
                />
              </div>
            ))}
        </div>
      )
    );
  }
}

AggregationFilters.propTypes = {
  onAggregationChange: PropTypes.func.isRequired,
  aggregations: PropTypes.instanceOf(Immutable.Map).isRequired,
  query: PropTypes.objectOf(PropTypes.any).isRequired,
  numberOfResults: PropTypes.number.isRequired,
};

export default AggregationFilters;
