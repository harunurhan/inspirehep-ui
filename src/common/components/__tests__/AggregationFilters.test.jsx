import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import AggregationFilters from '../AggregationFilters';
import AggregationFilter from '../AggregationFilter';

describe('AggregationFilters', () => {
  it('renders with all props set', () => {
    const aggregations = fromJS({
      agg1: {
        buckets: [
          {
            key: 'foo',
            doc_count: 1,
          },
        ],
        meta: {
          title: 'Aggregation 1',
          order: 1,
        },
      },
      agg2: {
        buckets: [
          {
            key: 'foo',
            doc_count: 1,
          },
        ],
        meta: {
          title: 'Aggregation 2',
          order: 2,
        },
      },
    });
    const query = { agg1: 'foo' };
    const wrapper = shallow(
      <AggregationFilters
        query={query}
        aggregations={aggregations}
        numberOfResults={2}
        onAggregationChange={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render aggregations with empty buckets', () => {
    const aggregations = fromJS({
      agg1: {
        buckets: [{}],
        meta: {
          title: 'Aggregation 1',
          order: 1,
        },
      },
      agg2: {
        buckets: [],
        meta: {
          title: 'Aggregation 2',
          order: 2,
        },
      },
    });
    const query = {};
    const wrapper = shallow(
      <AggregationFilters
        query={query}
        aggregations={aggregations}
        numberOfResults={2}
        onAggregationChange={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render aggregations when numberOfResults is 0', () => {
    const aggregations = fromJS({
      agg: {
        buckets: [
          {
            key: 'foo',
            doc_count: 0,
          },
        ],
        meta: {
          title: 'Jessica Jones',
          order: 1,
        },
      },
      emptyAgg: {
        buckets: [],
        meta: {
          title: 'Luke Cage',
          order: 2,
        },
      },
    });
    const query = {};
    const wrapper = shallow(
      <AggregationFilters
        query={query}
        aggregations={aggregations}
        numberOfResults={0}
        onAggregationChange={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with range aggregation (earliest_date)', () => {
    const aggregations = fromJS({
      earliest_date: {
        buckets: [
          {
            key: '2000',
            doc_count: 1,
          },
          {
            key: '2001',
            doc_count: 1,
          },
        ],
        meta: {
          title: 'Range Aggregation',
          order: 1,
        },
      },
    });
    const query = {};
    const wrapper = shallow(
      <AggregationFilters
        query={query}
        aggregations={aggregations}
        numberOfResults={2}
        onAggregationChange={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls onAggregationChange when aggregation is changed', () => {
    const aggregations = fromJS({
      agg: {
        buckets: [
          {
            key: 'foo',
            doc_count: 1,
          },
          {
            key: 'bar',
            doc_count: 2,
          },
          {
            key: 'uncool',
            doc_count: 3,
          },
        ],
        meta: {
          title: 'Aggregation',
          order: 1,
        },
      },
    });
    const query = {};
    const onAggregationChange = jest.fn();
    const wrapper = shallow(
      <AggregationFilters
        query={query}
        aggregations={aggregations}
        numberOfResults={2}
        onAggregationChange={onAggregationChange}
      />
    );
    const onAggregationFilterChange = wrapper
      .find(AggregationFilter)
      .prop('onChange');
    onAggregationFilterChange(['foo', 'bar']);
    expect(onAggregationChange).toBeCalledWith('agg', ['foo', 'bar']);
  });

  it('calls onAggregationChange when range aggregation (earliest_date) is changed', () => {
    const aggregations = fromJS({
      earliest_date: {
        buckets: [
          {
            key: '2000',
            doc_count: 1,
          },
          {
            key: '2001',
            doc_count: 1,
          },
        ],
        meta: {
          title: 'Range Aggregation',
          order: 1,
        },
      },
    });
    const query = {};
    const onAggregationChange = jest.fn();
    const wrapper = shallow(
      <AggregationFilters
        query={query}
        aggregations={aggregations}
        numberOfResults={2}
        onAggregationChange={onAggregationChange}
      />
    );
    const onAggregationFilterChange = wrapper
      .find(AggregationFilter)
      .prop('onChange');
    onAggregationFilterChange(['2000', '2001']);
    expect(onAggregationChange).toBeCalledWith('earliest_date', '2000--2001');
  });
});
