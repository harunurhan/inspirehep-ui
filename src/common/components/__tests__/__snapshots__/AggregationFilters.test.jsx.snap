// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AggregationFilters does not render aggregations when numberOfResults is 0 1`] = `""`;

exports[`AggregationFilters does not render aggregations with empty buckets 1`] = `
<div
  className="bg-white pa3"
>
  <div
    key="agg1"
  >
    <AggregationFilter
      buckets={
        Immutable.List [
          Immutable.Map {
          },
        ]
      }
      name="Aggregation 1"
      onChange={[Function]}
      range={false}
      selections={Array []}
    />
  </div>
</div>
`;

exports[`AggregationFilters renders with all props set 1`] = `
<div
  className="bg-white pa3"
>
  <div
    key="agg1"
  >
    <AggregationFilter
      buckets={
        Immutable.List [
          Immutable.Map {
            key: "foo",
            doc_count: 1,
          },
        ]
      }
      name="Aggregation 1"
      onChange={[Function]}
      range={false}
      selections={
        Array [
          "foo",
        ]
      }
    />
  </div>
  <div
    key="agg2"
  >
    <AggregationFilter
      buckets={
        Immutable.List [
          Immutable.Map {
            key: "foo",
            doc_count: 1,
          },
        ]
      }
      name="Aggregation 2"
      onChange={[Function]}
      range={false}
      selections={Array []}
    />
  </div>
</div>
`;

exports[`AggregationFilters renders with range aggregation (earliest_date) 1`] = `
<div
  className="bg-white pa3"
>
  <div
    key="earliest_date"
  >
    <AggregationFilter
      buckets={
        Immutable.List [
          Immutable.Map {
            key: "2000",
            doc_count: 1,
          },
          Immutable.Map {
            key: "2001",
            doc_count: 1,
          },
        ]
      }
      name="Range Aggregation"
      onChange={[Function]}
      range={true}
      selections={Array []}
    />
  </div>
</div>
`;
